### ISSUE TITLE
Adding fuzzing support for self-driving functionalities

### ISSUE THREAD

**Issue by jiayunhan** [ApolloAuto/apollo#5222]
I am planning to add the fuzzing support to the Apollo code base, which is controlled by a new build command `./apollo.sh build_fuzz_test` into the build script. This command will build the fuzzing test cases  whose source code are located in the `modules/tools/fuzz/` folder (currently covers several functionalities under `control/` and `prediction/` modules). 

As an example, considering the testing of the the `control` module, the fuzzer will generate `PadMessage`, `LocalizationEstimate`, `ADCTrajectory` and `Chassis` messages to feed into the control module, before it `ProduceCommand`. After each run, the input messages will be mutated based on the coverage feedback before fed again. This helps explore different states  of the program quickly, and will report problems (usually memory corruptions) with the help of `AddressSanitizer` and `UndefinedBehaviorSanitizer`. 

Fuzzing support is built on top of [libfuzzer](https://llvm.org/docs/LibFuzzer.html), which requires `clang`. I have add a `CROSSTOOL` for `clang` under the `tools/` folder for building the fuzz test. It relies on the `libprotobuf-mutator` to perform structure(protobuf)-aware mutations, which is hosted [here](https://github.com/BaiduXLab/libprotobuf-mutator) under BaiduXLab. The repo has been added to the `WORKSPACE.in`

More detail on how to utilize it, and how to extend it with more test cases can be found [here](https://github.com/BaiduXLab/libprotobuf-mutator). I would like to invite more people to contribute fuzz tests leveraging our structure-aware fuzzing support to improve Apollo. 

The [pull request](https://github.com/ApolloAuto/apollo/pull/5223) is open now.

**Comment by jiayunhan** [comment]
Prerequisite: Clang-6.0. Use the following command in the **docker** to install before running `build_fuzz_test`
```
sudo apt-add-repository "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-6.0 main"
wget -O - http://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
sudo apt-get update
sudo apt-get install clang-6.0 lldb-6.0 lld-6.0
sudo ln -s /usr/lib/x86_64-linux-gnu/libgfortran.so.3 /usr/lib/libgfortran.so
```

**Comment by natashadsouza** [comment]
Currently closing this issue as the resolving code has been merged into master.
